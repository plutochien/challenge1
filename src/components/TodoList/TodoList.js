import { sortBy } from 'lodash';
import PropTypes from 'prop-types';
import React, { Fragment, useEffect } from 'react';
import { connect } from 'react-redux';

import { fetchLocalData, toggleComplete } from '../../actions/todoAction';
import { FILTERS } from '../../util/constatns';
import Task from '../Task/Task';

const TodoList = ({ list, fetchLocalData, toggleComplete }) => {
  useEffect(() => {
    fetchLocalData();
  }, [fetchLocalData]);

  return (
    <Fragment>
      {list.map((task, index) => (
        <Task
          key={task.id}
          data={{ ...task, order: index + 1 }}
          toggleComplete={() => {
            toggleComplete(task.id);
          }}
        />
      ))}
    </Fragment>
  );
};

TodoList.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      isComplete: PropTypes.bool.isRequired,
      taskText: PropTypes.string.isRequired,
      priority: PropTypes.string.isRequired,
    }).isRequired,
  ).isRequired,
  fetchLocalData: PropTypes.func.isRequired,
  toggleComplete: PropTypes.func.isRequired,
};

const filterList = (list, filter) => {
  switch (filter) {
    case FILTERS.SHOW_ALL:
      return sortByPriority(list);
    case FILTERS.SHOW_INCOMPLETE:
      return sortByPriority(list.filter(todo => !todo.isComplete));
    case FILTERS.SHOW_COMPLETE:
      return sortByPriority(list.filter(todo => todo.isComplete));
    default:
      return sortByPriority(list);
  }
};

const sortByPriority = (list) => {
  return sortBy(list, ['priority', 'taskText']);
}

const mapStateToProps = ({ todos, filter }) => ({
  list: filterList(todos.list, filter),
});

const mapDispatchToProps = dispatch => ({
  toggleComplete: id => dispatch(toggleComplete(id)),
  fetchLocalData: () => dispatch(fetchLocalData()),
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoList);
