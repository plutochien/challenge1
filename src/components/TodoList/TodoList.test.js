import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import TodoList from './TodoList';
const mockStore = configureStore([]);

describe('TodoList Component', () => {
    let store;
    let component;
    beforeEach(() => {
        store = mockStore({
            todos: {
                list: [
                    {
                        id: 7,
                        taskText: 'task1',
                        isComplete: true,
                        priority: '0'
                    },
                    {
                        id: 6,
                        taskText: 'task2',
                        isComplete: false,
                        priority: '1'
                    },
                    {
                        id: 1,
                        taskText: 'task3',
                        isComplete: true,
                        priority: '2'
                    },
                ],
            },
            filter: 'SHOW_ALL',
        });
        store.dispatch = jest.fn();
        const listdata = [
                {
                    id: 7,
                    taskText: 'task1',
                    isComplete: true,
                    priority: '2'
                },
                {
                    id: 6,
                    taskText: 'task2',
                    isComplete: false,
                    priority:'1'
                },
                {
                    id: 1,
                    taskText: 'task3',
                    isComplete: true,
                    priority:'0'
                },
            ];
        component = renderer.create(
            <Provider store={store}>
                <TodoList listdata={listdata} fetchLocalData={jest.fn} toggleComplete={jest.fn}/>
            </Provider>,
        );
    });

    it('should render with given state from Redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

});
