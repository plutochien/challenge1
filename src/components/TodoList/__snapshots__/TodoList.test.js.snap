// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component should render with given state from Redux store 1`] = `
Array [
  <div>
    <li
      className="sc-bdfBwQ CGZeV"
    >
      <span
        className="sc-dlfnbm dhyx"
      >
        1
         
      </span>
      <span
        className="sc-gsTCUz gHDLyj"
      >
        task1
        <span
          className="sc-jSgupP cwbpyv"
        >
          Done
        </span>
      </span>
      <button
        className="sc-hKgILt bzvHwI"
        onClick={[Function]}
        size="40"
      >
        Undone
      </button>
      <button
        className="sc-eCssSg cnoREb"
        onClick={[Function]}
        size="40"
      >
        Delete
      </button>
    </li>
  </div>,
  <div>
    <li
      className="sc-bdfBwQ CGZeV"
    >
      <span
        className="sc-dlfnbm jAAmbI"
      >
        2
         
      </span>
      <span
        className="sc-gsTCUz byAzUf"
      >
        task2
        
      </span>
      <button
        className="sc-hKgILt fCEWoY"
        onClick={[Function]}
        size="40"
      >
        Done
      </button>
      <button
        className="sc-eCssSg cnoREb"
        onClick={[Function]}
        size="40"
      >
        Delete
      </button>
    </li>
  </div>,
  <div>
    <li
      className="sc-bdfBwQ CGZeV"
    >
      <span
        className="sc-dlfnbm cikqdF"
      >
        3
         
      </span>
      <span
        className="sc-gsTCUz cLFHbb"
      >
        task3
        <span
          className="sc-jSgupP cwbpyv"
        >
          Done
        </span>
      </span>
      <button
        className="sc-hKgILt bzvHwI"
        onClick={[Function]}
        size="40"
      >
        Undone
      </button>
      <button
        className="sc-eCssSg cnoREb"
        onClick={[Function]}
        size="40"
      >
        Delete
      </button>
    </li>
  </div>,
]
`;
